### ISSUE TITLE
Skip cell type annotation for cell line samples 

### ISSUE THREAD

**Issue by allyhawkins** [AlexsLemonade/scpca-nf#640]
**Is your feature request related to a problem? Please describe.**

We have some samples that are from cell lines rather than tumor samples. Although there may be different cell states, we don't expect to find very different cell types within a cancer cell line. Because of this, we should skip cell type annotation. 

**Describe the solution you'd like**

We will be adding a `is_cell_line` column to the sample metadata. Currently, the path to that file is input to the workflow. We don't actually do anything within nextflow with that file and it gets read in within the R script that generates the original SCE objects. This means that we aren't actually tracking any sample metadata throughout the workflow. 
https://github.com/AlexsLemonade/scpca-nf/blob/512d97be7aee8f90c9e0bb7547446d627f51cdfe/modules/sce-processing.nf#L242

We should change this so that we read in the sample metadata file (similar to how we read in the library metadata file) and create a sample channel. We can then grab the `is_cell_line` and `library_id` columns from the sample channel and join with the library metadata channel (`unfiltered_runs_ch`). This should be done in `main.nf`. 

Once we add `is_cell_line` to the `unfiltered_runs_ch`, it will become part of the `meta` object that we use throughout the workflow. Then we can use that to skip the `annotate_celltypes` workflow. 

**Additional context**

Originally, when we talked about this, I thought we read in the sample metadata as a channel, like we do with the library metadata. That's not the case since we just use the path to that file as input to one of the processes right now. However, I think this should still be fairly straightforward since this is the only column we need, and we can use the `library_id` column to join.

**Comment by sjspielman** [comment]
Hm I was all excited to start this today, but obviously we can't start this until wrapping up #751...! So instead, I'm adding the blocker.

**Comment by sjspielman** [comment]
I've started looking into this, and (surprise!) came across a bit of a hurdle. For posterity, copying in below some of the Nextflow Console code I was using to get the ball rolling. This code fails, not unexpectedly!, but here's the idea:

We need to first read in the sample metadata and grab sample id (to use for joining), and the cell line column. Then, we need to join with the library metadata. The challenge though is multiplex samples where there are multiple samples per library, and in `unfiltered_runs_ch` they are comma-separated. 

In order to get around this challenge, we have two choices:
- Embark on some ~extravagant nextflow wrangling and joining journey to correctly join this information together. For this, we should definitely get together for a technical design discussion.
- Modify `scpca-library-metadata.tsv` to also include the `is_cell_type` column which is currently only present in `scpca-sample-metadata.tsv`. This will result in some duplicate metadata, but it will also make our lives a lot easier here, such that no joining will need needed at all; we'd just be able to parse `is_cell_type` when creating `unfiltered_runs_ch` in the first place.

Tagging @allyhawkins @jashapiro for discussion!

```
 is_cell_line_ch = Channel.fromPath(sample_meta_path)
    .splitCsv(header: true, sep: '\t')
    .map{
      [
        sample_id: it.scpca_sample_id,
        is_cell_line: parseNA(it.is_cell_line).toBoolean() // FALSE -> false, NA -> false, TRUE -> true
      ]
    }


unfiltered_runs_ch = Channel.fromPath(library_meta_path)
    .splitCsv(header: true, sep: '\t')
    .map{[
      project_id: it.scpca_project_id, 
      sample_id: it.scpca_sample_id.split(";").sort().join(","), // this is how we parse when creating the meta map
      library_id: it.scpca_library_id 
     ]}
     // bring sample_id to the front for joining 
    .map{it.sample_id + it}
    // Fails, not surprisingly. In theory this should _not_ fail if there are only non-multiplex samples, but it still seems to fail (weird?) even if I filter out the multiplexed projects
    .join(is_cell_line_ch, by: 0, failOnMismatch: true, failOnDuplicate: true)
    .view()
```

**Comment by allyhawkins** [comment]
> Modify scpca-library-metadata.tsv to also include the is_cell_type column which is currently only present in scpca-sample-metadata.tsv. This will result in some duplicate metadata, but it will also make our lives a lot easier here, such that no joining will need needed at all; we'd just be able to parse is_cell_type when creating unfiltered_runs_ch in the first place.

I am totally on board with this plan.

**Comment by jashapiro** [comment]
In this case, I think you need `.combine()` instead of `.join()`, confusingly. `.join()` expects 1 and only 1 of each key: it will emit only the first instance. `.combine()` allows for more than one match and emits the cartesian product, so it is what you want when you expect something other than a 1:1 relationship. In other words, it is more the SQL `JOIN` equivalent.

**Comment by allyhawkins** [comment]
> In this case, I think you need .combine() instead of .join(), confusingly. .join() expects 1 and only 1 of each key in both channels: it will emit only the first instance. .combine() allows for more than one match and emits the cartesian product, so it is what you want when you expect something other than a 1:1 relationship. In other words, it is more the SQL JOIN equivalent.

I'm pretty sure this will still fail trying to join on sample ID's though because the sample ID listed in the sample metadata will be one sample per row, while in the library metadata has a comma separated list of sample ID's for multiplexed samples. So regardless we have to decide if we want to do some Nextflow wrangling to deal with joining the multiplex samples or just add it to the library metadata directly.

**Comment by jashapiro** [comment]
That is true, but I think we can actually handle things without `.combine` or `.join` to do something a bit different than creating a full extra dataset

I would start with the sample info table and then use `.filter()` to get just the samples which are cell lines. Then output that with `.toList()`

Then we can use `.branch{}` to pull out the ones that are present. 

Groovy syntax is odd, but an example:
```
sample_id = "test1,test2"
cell_lines = ["test4","test3"]

print( sample_id.split(",").collect{it in cell_lines}.any())
```

**Comment by jashapiro** [comment]
Just to complete the thought with a "full" example:
```
library_ch = Channel.of([library_id: "A", sample_id: "test1"],
                       [library_id: "B", sample_id: "test2"], 
                       [library_id: "C", sample_id: "test1,test2"], 
                       [library_id: "D", sample_id: "test3"])

cell_lines = Channel.of(["test1", true],["test2", true], ["test3", false] )
  .filter{it[1]}
  .map{it[0]}
  .toList()
  
branched = library_ch.branch{
  has_cell_line: it.sample_id.split(",").collect{it in cell_lines.getVal()}.any()
  other: true
 }
 

 branched.has_cell_line.view{"has cell line: $it.library_id"}
 

 branched.other.view{"no cell line: $it.library_id"}
 
```

**Comment by sjspielman** [comment]
Closed by #648
